gh-74929: PEP 667 general docs update (#119201)
python/cpython: Doc/library/functions.rst

   .. note::

      When ``exec`` gets two separate objects as *globals* and *locals*, the
      code will be executed as if it were embedded in a class definition. This
      means functions defined in the executed code will not be able to access
      variables assigned at the top level (as the "top level" variables are
      treated as class variables in a class definition).
      Passing a :class:`collections.ChainMap` instance as *globals* allows name
      lookups to be chained across multiple mappings without triggering this
      behaviour. Values assigned to top-level names in the executed code can be

>       retrieved by passing an empty dictionary as the first entry in the chain.

Author: This change on `main` triggered a conflict with the PR. While "Most users should just pass a *globals* argument and never *locals*" is technically true, it leaves the question of "What if I need to capture the names bound by the executed code?"  open, so I added a direct reference out to `collections.ChainMap` as a way to collect the writes without changing the code execution semantics.

I also added more details on the most common way that treating the code to be executed as class body code can cause problems.

Author: I also added a reference to this PR and comment from https://github.com/python/cpython/pull/119235 (the PR where this note originated)

Author: My change here wasn't correct, so I posted a follow-up PR to fix it: https://github.com/python/cpython/pull/119378

    In all of the above cases, each call to ``locals()`` in a given frame of
    execution will return the *same* mapping object. Changes made through

>     the mapping object returned from ``locals()`` will be visible as assigned,

Author: Since we have assignment statements rather than binding statements, I realised I should use different verbs in these docs.

Reviewer: I imagine that 'bind' is still used here and there in the docs, but changing here while editing is fine.  Either way, users must learn that the joining action happens in a target namespace rather than in either object.

Author: Yeah, we definitely do use "binding" and "bound" in various places. It just looked weird here once I saw it in context with the rest of the builtin docs rather than standing on its own in PEP 558 and the pre-beta-1 PR diff.

   .. versionchanged:: 3.13
      As part of :pep:`667`, calling `locals()` in a comprehension at module
      or class scope (including via `exec` or `eval`) once more behaves as if
      the comprehension were running as an independent nested function (i.e.
      the local variables from the containing scope are not included).

   .. versionchanged:: 3.12
      As part of :pep:`709`, calling `locals()` in a comprehension at module
      or class scope (including via `exec` or `eval`) produces an independent
      snapshot of the containing scope with the comprehension iteration
      variable(s) included.

   .. versionchanged:: 3.12
      As part of :pep:`709`, calling `locals()` in a comprehension as part of a
      function, generator, or coroutine returns the same `locals()` reference
      as calls in the containing scope.

   ..
      Possible simplification: change the 3.12 note to just "Changes to
      behavior in comprehensions as described in :pep:`709`", and do somethine
      similar for 3.13 and :pep:`667`. While the draft change notes are
      comprehensive, they're also almost entirely in the "Who is really going

>       to care?" category.

Author: I currently plan to follow through with this simplification before merging, but wanted to give others a chance to share their thoughts (I may also transfer some of this detail to the porting section in the What's New rather than dropping it entirely)

Author: The arguably useful parts of these notes has been included in the What's New Python API porting guide, so I went ahead and replaced them with the simplified version.

   :attr:`~object.__dict__` attributes (for example, classes use a
   :class:`types.MappingProxyType` to prevent direct dictionary updates).

   Without an argument, :func:`vars` acts like :func:`locals`.

   A :exc:`TypeError` exception is raised if an object is specified but
   it doesn't have a :attr:`~object.__dict__` attribute (for example, if
   its class defines the :attr:`~object.__slots__` attribute).

   .. versionchanged:: 3.13

>       The semantics of calling this function without an argument have been

Reviewer: Possibly
```suggestion
      The result of calling this function without an argument has been
```

Author: Suggestion accepted (along with changing "adjusted" to "updated")
